File: analyze-conflicts.ts
#!/usr/bin/env bun
import fs from 'fs';

// Our specification's key components
const ourSpec = {
  name: "Activity Template Flow via MCP Backend",
  components: [
    "repos/metabob-opencode/packages/opencode/src/session/template-loader.ts",
    "repos/metabob-opencode/packages/opencode/src/server/template-service-client.ts",
    "repos/metabob-opencode/packages/opencode/src/util/metabob.ts",
    "repos/metabob-opencode/packages/opencode/src/agent/agent.ts",
    "repos/metabob-opencode/packages/opencode/src/tool/activity.ts",
    "repos/metabob-rpc-api/server/routes/activity.py",
    "repos/metabob-rpc-api/server/actions/activity.py"
  ],
  requirements: [
    "TemplateLoader returns source='metabob' for backend templates",
    "No direct file access to .metabob/activities/*.json",
    "MetabobCLI has no local template writes",
    "Activity agent has search_activities, no impulse tools",
    "Memory agent has impulse tools",
    "TemplateServiceClient delegates to MetabobCLI"
  ]
};

// Load all validation results
const validationFiles = fs.readdirSync('./impulses')
  .filter(f => f.startsWith('validation-results-') && f.endsWith('.json'))
  .filter(f => f !== 'validation-results-activity-template-flow-via-mcp-backend.json');

console.log(`Found ${validationFiles.length} other validation results to analyze\n`);

const relatedSpecs: any[] = [];
const potentialConflicts: any[] = [];

validationFiles.forEach(file => {
  try {
    const content = JSON.parse(fs.readFileSync(`./impulses/${file}`, 'utf-8'));
    const specName = content.metadata?.specification || file.replace('validation-results-', '').replace('.json', '');
    
    // Check if this spec might touch our components
    const contentStr = JSON.stringify(content).toLowerCase();
    
    const touchesOurComponents = 
      contentStr.includes('template') ||
      contentStr.includes('activity') ||
      contentStr.includes('agent') ||
      contentStr.includes('mcp') ||
      contentStr.includes('metabob');
    
    if (touchesOurComponents) {
      relatedSpecs.push({
        name: specName,
        file,
        status: content.metadata?.overallStatus || 'UNKNOWN',
        tests: content.metadata?.totalTests || 0,
        passed: content.metadata?.passed || 0
      });
    }
  } catch (error) {
    // Skip invalid JSON
  }
});

console.log('═'.repeat(70));
console.log('Related Specifications');
console.log('═'.repeat(70));

relatedSpecs.forEach((spec, i) => {
  console.log(`\n${i + 1}. ${spec.name}`);
  console.log(`   Status: ${spec.status}`);
  console.log(`   Tests: ${spec.passed}/${spec.tests}`);
  console.log(`   File: ${spec.file}`);
});

console.log(`\n${'═'.repeat(70)}`);
console.log(`Total Related Specs: ${relatedSpecs.length}`);
console.log('═'.repeat(70));

// Analyze specific conflicts
console.log('\n\n🔍 Analyzing Potential Conflicts...\n');

// Check for architecture separation specs
const architectureSeparation = relatedSpecs.find(s => 
  s.name.includes('architecture-separation') || s.name.includes('complete-architecture')
);

if (architectureSeparation) {
  console.log(`✅ Found: ${architectureSeparation.name}`);
  console.log(`   This spec likely ALIGNS with our specification (both enforce separation)`);
  console.log(`   Status: ${architectureSeparation.status}`);
}

// Check for MCP data flow specs
const mcpDataFlow = relatedSpecs.filter(s => 
  s.name.includes('mcp-data-flow') || s.name.includes('mcp')
);

if (mcpDataFlow.length > 0) {
  console.log(`\n✅ Found ${mcpDataFlow.length} MCP data flow spec(s):`);
  mcpDataFlow.forEach(s => {
    console.log(`   - ${s.name} (${s.status})`);
  });
  console.log(`   These specs likely ALIGN with our MCP backend flow requirement`);
}

// Check for activity-related specs
const activitySpecs = relatedSpecs.filter(s => 
  s.name.includes('activity') && !s.name.includes('template-flow')
);

if (activitySpecs.length > 0) {
  console.log(`\n⚠️  Found ${activitySpecs.length} other activity-related spec(s):`);
  activitySpecs.forEach(s => {
    console.log(`   - ${s.name} (${s.status})`);
  });
  console.log(`   Need to check if these specs have conflicting requirements`);
}

// Check for bootstrap template specs
const bootstrapSpecs = relatedSpecs.filter(s => 
  s.name.includes('bootstrap')
);

if (bootstrapSpecs.length > 0) {
  console.log(`\n✅ Found ${bootstrapSpecs.length} bootstrap template spec(s):`);
  bootstrapSpecs.forEach(s => {
    console.log(`   - ${s.name} (${s.status})`);
  });
  console.log(`   These specs define bootstrap fallback behavior (complements our spec)`);
}

// Write analysis to file
const analysis = {
  specificationName: ourSpec.name,
  analyzedDate: new Date().toISOString(),
  totalRelatedSpecs: relatedSpecs.length,
  relatedSpecs: relatedSpecs.map(s => ({
    name: s.name,
    status: s.status,
    relationship: 
      s.name.includes('architecture-separation') ? 'ALIGNS' :
      s.name.includes('mcp') ? 'ALIGNS' :
      s.name.includes('bootstrap') ? 'COMPLEMENTS' :
      'UNKNOWN'
  })),
  potentialConflicts: [],
  sharedComponents: ourSpec.components
};

fs.writeFileSync('./conflict-analysis-preliminary.json', JSON.stringify(analysis, null, 2));
console.log('\n\n✅ Preliminary analysis saved to conflict-analysis-preliminary.json');

File: analyze-conflicts.py
#!/usr/bin/env python3
"""
Analyze conflicts between user-authentication-login-flow-fix and other specifications
"""

import json
import os
from pathlib import Path

# Our specification
CURRENT_SPEC = "user-authentication-login-flow-fix"

# Files modified by our spec
OUR_FILES = {
    "repos/metabob-rpc-api/server/db/operations/user_ops.py": {
        "changes": ["User ID format changed from hyphens to underscores"],
        "components": ["create_user", "user ID generation"]
    },
    "repos/metabob-rpc-api/server/routes/cloud_auth.py": {
        "changes": ["Fixed SurrealDB query result parsing", "Added debug logging"],
        "components": ["login endpoint", "query result parsing"]
    },
    "scripts/init-surrealdb-devbob-schema-v2.sql": {
        "changes": ["Added users, organizations, user_organizations, refresh_tokens tables"],
        "components": ["database schema", "authentication tables"]
    }
}

def find_related_specs():
    """Find other specifications that touch related components"""
    related = []
    
    # Scan validation results
    impulse_dir = Path("impulses")
    if not impulse_dir.exists():
        print(f"Warning: {impulse_dir} not found")
        return related
    
    for file in impulse_dir.glob("validation-results-*.json"):
        try:
            with open(file, 'r') as f:
                data = json.load(f)
            
            spec_name = data.get('metadata', {}).get('specificationName') or data.get('specificationName')
            if not spec_name or spec_name == CURRENT_SPEC:
                continue
            
            # Check if this spec touches our files
            touches_our_files = False
            affected_components = []
            
            # Check for overlaps in content
            content_str = json.dumps(data)
            
            if 'user_ops' in content_str or 'create_user' in content_str:
                touches_our_files = True
                affected_components.append('user_ops.py')
            
            if 'cloud_auth' in content_str or 'login' in content_str:
                touches_our_files = True
                affected_components.append('cloud_auth.py')
            
            if 'schema' in content_str or 'surrealdb' in content_str.lower():
                touches_our_files = True
                affected_components.append('SurrealDB schema')
            
            if touches_our_files:
                related.append({
                    'spec': spec_name,
                    'file': str(file.name),
                    'components': affected_components
                })
        
        except Exception as e:
            print(f"Error reading {file}: {e}")
    
    return related

def analyze_conflicts(related_specs):
    """Analyze potential conflicts"""
    conflicts = []
    
    # Known conflict patterns
    for spec_info in related_specs:
        spec = spec_info['spec']
        components = spec_info['components']
        
        # Check for SurrealDB-related conflicts
        if 'SurrealDB schema' in components:
            conflicts.append({
                'type': 'SCHEMA_OVERLAP',
                'spec1': CURRENT_SPEC,
                'spec2': spec,
                'sharedComponent': 'scripts/init-surrealdb-devbob-schema-v2.sql',
                'description': f'{spec} may have also modified SurrealDB schema. Need to verify schema changes are compatible.',
                'severity': 'MEDIUM',
                'resolution': 'Review both schema files and merge changes if needed'
            })
        
        # Check for authentication-related conflicts
        if 'cloud_auth.py' in components:
            conflicts.append({
                'type': 'CODE_OVERLAP',
                'spec1': CURRENT_SPEC,
                'spec2': spec,
                'sharedComponent': 'repos/metabob-rpc-api/server/routes/cloud_auth.py',
                'description': f'{spec} may have modified cloud_auth.py. Our query parsing fix could conflict.',
                'severity': 'HIGH',
                'resolution': 'Review both implementations and ensure query parsing logic is consistent'
            })
        
        # Check for user creation conflicts
        if 'user_ops.py' in components:
            conflicts.append({
                'type': 'CODE_OVERLAP',
                'spec1': CURRENT_SPEC,
                'spec2': spec,
                'sharedComponent': 'repos/metabob-rpc-api/server/db/operations/user_ops.py',
                'description': f'{spec} may have modified user_ops.py. Our user ID format change could conflict.',
                'severity': 'HIGH',
                'resolution': 'Ensure user ID format (underscores) is consistent across all specs'
            })
    
    return conflicts

def check_deployment_conflicts():
    """Check for deployment-related conflicts"""
    deployment_conflicts = []
    
    # Our spec requires RPC API rebuild
    # Check if other recent specs also require this
    deployment_conflicts.append({
        'type': 'DEPLOYMENT_DEPENDENCY',
        'issue': 'Multiple specs require RPC API rebuild',
        'specs': [CURRENT_SPEC, 'surrealdb-authentication-fix-and-dashboard-live-test'],
        'description': 'Both specs modify RPC API code and require Docker rebuild/redeploy',
        'severity': 'LOW',
        'resolution': 'Rebuild RPC API image once with all changes combined'
    })
    
    return deployment_conflicts

def main():
    print("=" * 70)
    print("CONFLICT ANALYSIS: user-authentication-login-flow-fix")
    print("=" * 70)
    print()
    
    # Find related specifications
    print("Finding related specifications...")
    related_specs = find_related_specs()
    
    print(f"Found {len(related_specs)} related specifications:")
    for spec_info in related_specs:
        print(f"  - {spec_info['spec']}")
        print(f"    Components: {', '.join(spec_info['components'])}")
    print()
    
    # Analyze conflicts
    print("Analyzing conflicts...")
    conflicts = analyze_conflicts(related_specs)
    deployment_conflicts = check_deployment_conflicts()
    
    all_conflicts = conflicts + deployment_conflicts
    
    print(f"Found {len(all_conflicts)} potential conflicts:")
    for i, conflict in enumerate(all_conflicts, 1):
        print(f"\n{i}. {conflict['type']} ({conflict['severity']} severity)")
        print(f"   Component: {conflict.get('sharedComponent', conflict.get('issue'))}")
        print(f"   Description: {conflict['description']}")
        print(f"   Resolution: {conflict['resolution']}")
    
    # Generate summary
    summary = {
        'specificationName': CURRENT_SPEC,
        'otherSpecifications': [s['spec'] for s in related_specs],
        'totalConflicts': len(all_conflicts),
        'highSeverity': len([c for c in all_conflicts if c['severity'] == 'HIGH']),
        'mediumSeverity': len([c for c in all_conflicts if c['severity'] == 'MEDIUM']),
        'lowSeverity': len([c for c in all_conflicts if c['severity'] == 'LOW']),
        'conflicts': all_conflicts,
        'sharedComponents': [
            {
                'component': file,
                'affectedBySpecs': [CURRENT_SPEC, 'surrealdb-authentication-fix-and-dashboard-live-test'],
                'recommendation': f"Review both implementations of {file}"
            }
            for file in OUR_FILES.keys()
        ],
        'conflictImpulseId': f'conflict-analysis-{CURRENT_SPEC}'
    }
    
    # Save results
    output_file = f'CONFLICT_ANALYSIS_{CURRENT_SPEC}.json'
    with open(output_file, 'w') as f:
        json.dump(summary, f, indent=2)
    
    print()
    print("=" * 70)
    print(f"SUMMARY: {len(all_conflicts)} conflicts detected")
    print(f"  HIGH severity: {summary['highSeverity']}")
    print(f"  MEDIUM severity: {summary['mediumSeverity']}")
    print(f"  LOW severity: {summary['lowSeverity']}")
    print()
    print(f"Results saved to: {output_file}")
    print("=" * 70)
    
    return 0 if len(all_conflicts) == 0 else 1

if __name__ == '__main__':
    import sys
    sys.exit(main())
